### ISSUE TITLE
V4 pre-signed url gets SignatureDoesNotMatch error

### ISSUE THREAD

**Issue by ksmandersen** [JustinM1/S3SignerAWS#10]
I am tried to generate a valid V4 pre-signed url for uploading an image to S3. 

I have checked and double-checked this:

- I have a valid set of accessKey and secretKey
- The IAM user for the keys has PutObject & PutObjectAcl permissions for the correct bucket
- The bucket's ACL has "Any authenticated AWS user" Read, Write access
- My bucket name does not contain a ``-`` or a ``.``

Here is my code:
```swift
let path = generateUrlForAsset(withName: assetName, fileExtension: validated.fileExtension)
let requestHeaders = ["x-amz-acl": "public-read"]
let preSigned = try s3Signer.presignedURLV4(httpMethod: .put, urlString: path, expiration: .thirtyMinutes, headers: requestHeaders)
```
the ``generateUrlForAsset`` function generates a url like ``https://s3.amazonaws.com/mybucketname/originals/5FBD0704-882B-4A8A-B931-8DB03D8DECBC.png``

I've tried using the signed url both with a web form but also curl like so:

```bash
curl "https://s3.amazonaws.com/mybucketname/originals/5FBD0704-882B-4A8A-B931-8DB03D8DECBC.png?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=MYS3ISSUEDACCESSKEY%2F20170629%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20170629T125701Z&X-Amz-Expires=1800&X-Amz-SignedHeaders=host;x-amz-acl&X-Amz-Signature=68c7d70b876a581bb581d23b159c2161dcc4f76343468dc5d805813ef1724b9c" --upload-file test.png -H "x-amz-acl: public-read"
```

I am out of options for debugging so far. Do you have any idea what I might be doing wrong?

**Update 1:**
I studied the generated url further and I saw that the query parameter for signed headers seperate the headers by ; which doesn't seem like a valid query parameter value to mee

X-Amz-SignedHeaders=host;x-amz-acl

**Update 2:**
I remove the x-amz-acl header from the list of headers and the url works now.

**Update 3:**
Escaping the ``;`` is unfortunately not enough. 

I tried generating a url with the native aws-sdk for node.js and it generates the same url where the ``;`` is escaped, but it also appends ``&x-amz-acl=public-read``. This URL however works.

**Comment by ksmandersen** [comment]
I studied the generated url further and I saw that the query parameter for signed headers seperate the headers by ``;`` which doesn't seem like a valid query parameter value to mee

``X-Amz-SignedHeaders=host;x-amz-acl``

**Comment by ksmandersen** [comment]
I remove the x-amz-acl header from the list of headers and the url works now.

**Comment by JustinM1** [comment]
hey @ksmandersen, thanks for reporting the issue you were having. Were you able to get it working properly with your fork?

**Comment by ksmandersen** [comment]
@JustinM1 I wasn't able to get it to work with my fork either. I am not sure exactly what the issue is. I only know for certain that it only occurs when adding additional headers to the pre-sign request.

**Comment by ksmandersen** [comment]
@JustinM1 Adding no headers generates a valid url. Adding any kind of header generates a url with an invalid signature.

**Comment by JustinM1** [comment]
@ksmandersen alright thanks, I'll try to look into it as soon as possible.

**Comment by JustinM1** [comment]
@ksmandersen I was able to make some updates to the project, cleaned a lot of things up, and added tests from the AWS Test Suite and the S3 examples. Unfortunately, I have only been able to find one example using V4 pre-signed URL and it doesn't add extra headers.

https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-query-string-auth.html#query-string-auth-v4-signing-example

Without having a chance to test with a real request, I believe it was an encoding issue. I have updated how headers are signed for V4 pre-signed URL requests to include percentEncoding according to characters AWS specifies. If you get a chance to try it out, let me know. 

There are some minor breaking changes when you update to 3.0.
